File: prototype/rails/src/app/actions/orders/finalize_order_test.rb
require "commerce_test_case"

module Orders
  class FinalizeOrderTest < CommerceTestCase
    def test_an_approved_card_pays_the_order
      order = finalize(order_for(customer, listing(seller)), APPROVED_CARD)

      assert_equal Domain::Orders::OrderStatus::PAID, order.status
      assert_equal moment("2026-08-20 10:00:00"), order.finalized_at
    end

    def test_an_approved_card_records_the_payment
      order = finalize(order_for(customer, listing(seller)), APPROVED_CARD)

      payment = order.payments.sole
      assert_equal Domain::Payments::PaymentStatus::APPROVED, payment.status
      assert_equal 45_000, payment.amount_cents
      assert_equal "4242", payment.card_last_four
      assert_nil payment.decline_reason
    end

    def test_a_paid_order_holds_the_seller_net_in_escrow
      shop = seller
      order = finalize(order_for(customer, listing(shop)), APPROVED_CARD)

      entry = LedgerEntry.sole
      assert_equal Domain::Escrow::LedgerEntryType::HELD, entry.entry_type
      assert_equal 40_500, entry.amount_cents
      assert_equal shop.id, entry.seller_id
      assert_equal order.fulfillments.sole.id, entry.fulfillment_id
    end

    def test_a_paid_order_holds_one_amount_per_seller
      order = order_for(customer, listing(seller("Blue Kiln Studio")), listing(seller("Rye Press"), price_cents: 10_000))

      finalize(order, APPROVED_CARD)

      assert_equal [9000, 40_500], LedgerEntry.order(:amount_cents).pluck(:amount_cents)
    end

    def test_a_paid_order_tells_each_seller_their_item_sold
      shop = seller

      finalize(order_for(customer, listing(shop)), APPROVED_CARD)

      notification = Notification.sole
      assert_equal shop.id, notification.seller_id
      assert_equal "Item sold", notification.subject
      assert_includes notification.body, "$405.00"
    end

    def test_a_declined_card_fails_the_payment
      order = finalize(order_for(customer, listing(seller)), DECLINED_CARD)

      assert_equal Domain::Orders::OrderStatus::PAYMENT_FAILED, order.status
      assert_nil order.finalized_at
      assert_equal Domain::Payments::DeclineReason::GENERIC_DECLINE, order.payments.sole.decline_reason
    end

    def test_a_declined_card_puts_the_stock_back_on_the_storefront
      art = listing(seller, quantity: 1)

      finalize(order_for(customer, art), DECLINED_CARD)

      art.reload
      assert_equal 1, art.quantity
      assert_equal Domain::Listings::ListingStatus::FOR_SALE, art.status
    end

    def test_a_declined_card_holds_nothing_and_tells_nobody
      finalize(order_for(customer, listing(seller)), DECLINED_CARD)

      assert_equal 0, LedgerEntry.count
      assert_equal 0, Notification.count
    end

    def test_a_retry_with_a_good_card_pays_the_order_and_takes_the_stock_again
      art = listing(seller, quantity: 1)
      order = order_for(customer, art)
      finalize(order, DECLINED_CARD)

      finalize(order, APPROVED_CARD, at: "2026-08-20 10:05:00")

      art.reload
      assert_equal Domain::Orders::OrderStatus::PAID, order.status
      assert_equal 0, art.quantity
      assert_equal Domain::Listings::ListingStatus::SOLD, art.status
      assert_equal 2, order.payments.count
      assert_equal 40_500, LedgerEntry.sole.amount_cents
    end

    def test_a_retry_that_is_declined_again_leaves_the_stock_on_the_storefront
      art = listing(seller, quantity: 1)
      order = order_for(customer, art)
      finalize(order, DECLINED_CARD)

      finalize(order, UNFUNDED_CARD, at: "2026-08-20 10:05:00")

      art.reload
      assert_equal Domain::Orders::OrderStatus::PAYMENT_FAILED, order.status
      assert_equal 1, art.quantity
      assert_equal Domain::Listings::ListingStatus::FOR_SALE, art.status
      assert_equal Domain::Payments::DeclineReason::INSUFFICIENT_FUNDS, order.payments.last.decline_reason
    end

    def test_it_refuses_to_charge_an_order_that_is_already_paid
      order = finalize(order_for(customer, listing(seller)), APPROVED_CARD)

      assert_raises(Domain::TransitionError) { finalize(order, APPROVED_CARD, at: "2026-08-20 10:05:00") }
    end

    def test_it_refuses_to_charge_an_order_that_has_not_been_verified
      order = order_for(anonymous_customer, listing(seller))

      assert_raises(Domain::TransitionError) { finalize(order, APPROVED_CARD) }
    end

    private

    def finalize(order, card_number, at: "2026-08-20 10:00:00")
      FinalizeOrder.new.call(order: order, card_number: card_number, now: moment(at))
    end
  end
end

File: prototype/php/src/app/Actions/Orders/FinalizeOrderTest.php
<?php

declare(strict_types=1);

namespace App\Actions\Orders;

use App\Domain\DomainRuleViolation;
use App\Domain\Escrow\LedgerEntryType;
use App\Domain\Listings\ListingStatus;
use App\Domain\Orders\OrderStatus;
use App\Domain\Payments\DeclineReason;
use App\Domain\Payments\PaymentStatus;
use App\Models\CustomerBlock;
use App\Models\LedgerEntry;
use App\Notifications\ItemSold;
use DomainException;
use Illuminate\Support\Facades\Notification;

it('pays the order with an approved card', function (): void {
    $order = $this->orderFor($this->verifiedCustomer(), $this->listing($this->seller(), ['price_cents' => 45000]));

    $order = app(FinalizeOrder::class)($order, '4242 4242 4242 4242', $this->moment('2026-08-20 10:00:00'));

    expect($order->status)->toBe(OrderStatus::Paid)
        ->and($order->finalized_at?->format('Y-m-d H:i:s'))->toBe('2026-08-20 10:00:00');
});

it('records the payment for an approved card', function (): void {
    $order = $this->orderFor($this->verifiedCustomer(), $this->listing($this->seller(), ['price_cents' => 45000]));

    app(FinalizeOrder::class)($order, '4242 4242 4242 4242', $this->moment('2026-08-20 10:00:00'));

    $payment = $order->payments()->sole();

    expect($payment->status)->toBe(PaymentStatus::Approved)
        ->and($payment->amount_cents)->toBe(45000)
        ->and($payment->card_last_four)->toBe('4242')
        ->and($payment->decline_reason)->toBeNull();
});

it('holds the seller net in escrow for a paid order', function (): void {
    $seller = $this->seller();
    $order = $this->orderFor($this->verifiedCustomer(), $this->listing($seller, ['price_cents' => 45000]));

    app(FinalizeOrder::class)($order, '4242 4242 4242 4242', $this->moment('2026-08-20 10:00:00'));

    $entry = LedgerEntry::query()->sole();

    expect($entry->type)->toBe(LedgerEntryType::Held)
        ->and($entry->amount_cents)->toBe(40500)
        ->and($entry->seller_id)->toBe($seller->id)
        ->and($entry->fulfillment_id)->toBe($order->fulfillments()->sole()->id);
});

it('holds one amount per seller on a paid order', function (): void {
    $order = $this->orderFor(
        $this->verifiedCustomer(),
        $this->listing($this->seller('Blue Kiln Studio'), ['price_cents' => 45000]),
        $this->listing($this->seller('Rye Press'), ['price_cents' => 10000]),
    );

    app(FinalizeOrder::class)($order, '4242 4242 4242 4242', $this->moment('2026-08-20 10:00:00'));

    expect(LedgerEntry::query()->orderBy('amount_cents')->pluck('amount_cents')->all())->toBe([9000, 40500]);
});

it('tells each seller their item sold on a paid order', function (): void {
    Notification::fake();
    $seller = $this->seller();
    $order = $this->orderFor($this->verifiedCustomer(), $this->listing($seller, ['price_cents' => 45000]));

    app(FinalizeOrder::class)($order, '4242 4242 4242 4242', $this->moment('2026-08-20 10:00:00'));

    Notification::assertSentTo(
        $seller,
        ItemSold::class,
        fn (ItemSold $notification): bool => str_contains($notification->toArray($seller)['body'], '$405.00'),
    );
});

it('fails the payment for a declined card', function (): void {
    $order = $this->orderFor($this->verifiedCustomer(), $this->listing($this->seller(), ['price_cents' => 45000]));

    $order = app(FinalizeOrder::class)($order, '4000 0000 0000 0002', $this->moment('2026-08-20 10:00:00'));

    expect($order->status)->toBe(OrderStatus::PaymentFailed)
        ->and($order->finalized_at)->toBeNull()
        ->and($order->payments()->sole()->decline_reason)->toBe(DeclineReason::GenericDecline);
});

it('puts the stock back on the storefront for a declined card', function (): void {
    $listing = $this->listing($this->seller(), ['price_cents' => 45000, 'quantity' => 1]);
    $order = $this->orderFor($this->verifiedCustomer(), $listing);

    app(FinalizeOrder::class)($order, '4000 0000 0000 0002', $this->moment('2026-08-20 10:00:00'));

    expect($listing->refresh()->quantity)->toBe(1)
        ->and($listing->status)->toBe(ListingStatus::ForSale);
});

it('holds nothing and tells nobody for a declined card', function (): void {
    Notification::fake();
    $order = $this->orderFor($this->verifiedCustomer(), $this->listing($this->seller(), ['price_cents' => 45000]));

    app(FinalizeOrder::class)($order, '4000 0000 0000 0002', $this->moment('2026-08-20 10:00:00'));

    expect(LedgerEntry::query()->count())->toBe(0);
    Notification::assertNothingSent();
});

it('pays the order and takes the stock again on a retry with a good card', function (): void {
    $listing = $this->listing($this->seller(), ['price_cents' => 45000, 'quantity' => 1]);
    $order = $this->orderFor($this->verifiedCustomer(), $listing);
    $finalizeOrder = app(FinalizeOrder::class);
    $finalizeOrder($order, '4000 0000 0000 0002', $this->moment('2026-08-20 10:00:00'));

    $order = $finalizeOrder($order, '4242 4242 4242 4242', $this->moment('2026-08-20 10:05:00'));

    expect($order->status)->toBe(OrderStatus::Paid)
        ->and($order->payments()->count())->toBe(2);
    expect($listing->refresh()->quantity)->toBe(0)
        ->and($listing->status)->toBe(ListingStatus::Sold);
    expect(LedgerEntry::query()->sole()->amount_cents)->toBe(40500);
});

it('refuses a blocked customer', function (): void {
    $customer = $this->verifiedCustomer();
    $order = $this->orderFor($customer, $this->listing($this->seller(), ['price_cents' => 45000]));
    CustomerBlock::factory()->create(['customer_id' => $customer->id, 'reason' => 'Chargeback fraud.']);

    $finalize = fn () => app(FinalizeOrder::class)($order, '4242 4242 4242 4242', $this->moment('2026-08-20 10:00:00'));

    expect($finalize)->toThrow(DomainRuleViolation::class, 'Chargeback fraud.')
        ->and($order->refresh()->payments()->count())->toBe(0);
});

it('refuses to charge an order that is already paid', function (): void {
    $order = $this->orderFor($this->verifiedCustomer(), $this->listing($this->seller(), ['price_cents' => 45000]));
    $finalizeOrder = app(FinalizeOrder::class);
    $order = $finalizeOrder($order, '4242 4242 4242 4242', $this->moment('2026-08-20 10:00:00'));

    $finalizeOrder($order, '4242 4242 4242 4242', $this->moment('2026-08-20 10:05:00'));
})->throws(DomainException::class);
